metadata-ingestion/src/datahub/utilities/file_backed_collections.py

```diff
@@ -29,56 +36,70 @@
+    a context manager interface.
     """
 
-    _ref_count: Dict[pathlib.Path, int] = field(default_factory=dict)
-    _sqlite_connection_cache: Dict[pathlib.Path, sqlite3.Connection] = field(
-        default_factory=dict
-    )
-
-    def get_connection(self, filename: pathlib.Path) -> sqlite3.Connection:
-        if filename not in self._ref_count:
-            conn = sqlite3.connect(filename, isolation_level=None)
-
-            # These settings are optimized for performance.
-            # See https://www.sqlite.org/pragma.html for more information.
-            # Because we're only using these dbs to offload data from memory, we don't need
-            # to worry about data integrity too much.
-            conn.execute('PRAGMA locking_mode = "EXCLUSIVE"')
-            conn.execute('PRAGMA synchronous = "OFF"')
-            conn.execute('PRAGMA journal_mode = "MEMORY"')
-            conn.execute(f"PRAGMA journal_size_limit = {100 * 1024 * 1024}")  # 100MB
-
-            self._ref_count[filename] = 0
-            self._sqlite_connection_cache[filename] = conn
-
-        self._ref_count[filename] += 1
-        return self._sqlite_connection_cache[filename]
+    conn: sqlite3.Connection
+    filename: pathlib.Path
+    _directory: Optional[tempfile.TemporaryDirectory] = None
+
+    def __init__(self, filename: Optional[pathlib.Path] = None):
+        # Warning: If filename is provided, the file will not be automatically cleaned up
+        if not filename:
+            self._directory = tempfile.TemporaryDirectory()
+            filename = pathlib.Path(self._directory.name) / _DEFAULT_FILE_NAME
+
+        self.conn = sqlite3.connect(filename, isolation_level=None)
+        self.filename = filename
+
+        # These settings are optimized for performance.
+        # See https://www.sqlite.org/pragma.html for more information.
+        # Because we're only using these dbs to offload data from memory, we don't need
+        # to worry about data integrity too much.
+        self.conn.execute('PRAGMA locking_mode = "EXCLUSIVE"')
+        self.conn.execute('PRAGMA synchronous = "OFF"')
+        self.conn.execute('PRAGMA journal_mode = "MEMORY"')
+        self.conn.execute(f"PRAGMA journal_size_limit = {100 * 1024 * 1024}")  # 100MB
+
+    def execute(
+        self, sql: str, parameters: Dict[str, Any] | Sequence[Any] = ()
+    ) -> sqlite3.Cursor:
+        logger.debug(f"Executing <{sql}> ({parameters})")
+        return self.conn.execute(sql, parameters)
+
+    def executemany(
+        self, sql: str, parameters: Dict[str, Any] | Sequence[Any] = ()
+    ) -> sqlite3.Cursor:
+        logger.debug(f"Executing many <{sql}> ({parameters})")
+        return self.conn.executemany(sql, parameters)
 
-    def drop_connection(self, filename: pathlib.Path) -> None:
-        self._ref_count[filename] -= 1
+    def close(self) -> None:
+        self.conn.close()
+        if self._directory:
+            self._directory.cleanup()
```

Comment:
same as below - if the call close manually / via the contextmanager and then the garbage collector calls `__del__`, then we'll call cleanup twice